pinot-core/src/main/java/com/linkedin/pinot/core/startree/StarTreeDataTableOptimized.java

```diff
@@ -0,0 +1,227 @@
+    Collections.sort(idList, comparator);
+    int[] currentPositions = new int[length];
+    int[] indexToRecordIdMapping = new int[length];
+
+    byte[] buf1 = new byte[totalSizeInBytes];
+    byte[] buf2 = new byte[totalSizeInBytes];
+
+    for (int i = 0; i < length; i++) {
+      currentPositions[i] = i;
+      indexToRecordIdMapping[i] = i;
+    }
+
+    for (int i = 0; i < length; i++) {
+      int thisRecordId = indexToRecordIdMapping[i];
+      int thisRecordIdPos = currentPositions[thisRecordId];
+
+      int thatRecordId = idList.get(i);
+      int thatRecordIdPos = currentPositions[thatRecordId];
+
+      // swap the buffers
+      mappedByteBuffer.copyTo(startOffset + thisRecordIdPos * totalSizeInBytes, buf1, 0, totalSizeInBytes);
+      mappedByteBuffer.copyTo(startOffset + thatRecordIdPos * totalSizeInBytes, buf2, 0, totalSizeInBytes);
+      mappedByteBuffer.readFrom(buf2, 0, startOffset + thisRecordIdPos * totalSizeInBytes, totalSizeInBytes);
+      mappedByteBuffer.readFrom(buf1, 0, startOffset + thatRecordIdPos * totalSizeInBytes, totalSizeInBytes);
+
+      // update the positions
+      indexToRecordIdMapping[i] = thatRecordId;
+      indexToRecordIdMapping[thatRecordIdPos] = thisRecordId;
+
+      currentPositions[thatRecordId] = i;
+      currentPositions[thisRecordId] = thatRecordIdPos;
+    }
+  }
+
+  /**
+   * Close the mappedByteBuffer
+   */
+  public void close() {
+    if (mappedByteBuffer != null) {
+      mappedByteBuffer.flush();
+      try {
+        mappedByteBuffer.close();
+      } catch (IOException e) {
+        LOGGER.error("Exception caught while trying to close byte-buffer", e);
+      }
+    }
+  }
+
+  /**
+   * @param startDocId inclusive
+   * @param endDocId exclusive
+   * @param colIndex
+   * @return start, end for each value. inclusive start, exclusive end
+   */
+  public Map<Integer, IntPair> groupByIntColumnCount(int startDocId, int endDocId, Integer colIndex) {
+    int length = endDocId - startDocId;
+    Map<Integer, IntPair> rangeMap = new LinkedHashMap<>();
+    final int startOffset = startDocId * totalSizeInBytes;
+
+    int prevValue = -1;
+    int prevStart = 0;
+    byte[] dimBuff = new byte[dimensionSizeInBytes];
+
+    for (int i = 0; i < length; i++) {
+      mappedByteBuffer.copyTo(startOffset + (i * totalSizeInBytes), dimBuff, 0, dimensionSizeInBytes);
+      int value = ByteBuffer.wrap(dimBuff).asIntBuffer().get(colIndex);
+      if (prevValue != -1 && prevValue != value) {
+        rangeMap.put(prevValue, new IntPair(startDocId + prevStart, startDocId + i));
+        prevStart = i;
+      }
+      prevValue = value;
+    }
+
+    rangeMap.put(prevValue, new IntPair(startDocId + prevStart, endDocId));
+    return rangeMap;
+  }
+
+  /**
+   * Returns iterator over dimension and metric byte buffers.
+   *
+   * @param startDocId
+   * @param endDocId
+   * @return
+   * @throws IOException
+   */
+  public Iterator<Pair<byte[], byte[]>> iterator(int startDocId, int endDocId)
```

Comment:
this method is re mapping the buffer. Either remove this method or don't re-map the file